pkg/certifier/datadog_malware/datadog_malware.go

```diff
@@ -0,0 +1,241 @@
+	httpClient    *http.Client
+	npmData       MaliciousPackages
+	pypiData      MaliciousPackages
+	assemblerFunc assemblerFuncType
+}
+
+// CertifierOption defines functional options for the certifier
+type CertifierOption func(*datadogMalwareCertifier)
+
+// WithHTTPClient allows overriding the default HTTP client
+func WithHTTPClient(client *http.Client) CertifierOption {
+	return func(d *datadogMalwareCertifier) {
+		d.httpClient = client
+	}
+}
+
+// NewDatadogMalwareCertifier initializes the Datadog Malicious Software Packages certifier
+func NewDatadogMalwareCertifier(ctx context.Context, assemblerFunc assemblerFuncType, opts ...CertifierOption) (certifier.Certifier, error) {
+	limiter := rate.NewLimiter(rate.Every(rateLimitInterval), rateLimit)
+	transport := clients.NewRateLimitedTransport(version.UATransport, limiter)
+	defaultClient := &http.Client{Transport: transport}
+
+	d := &datadogMalwareCertifier{
+		httpClient:    defaultClient,
+		assemblerFunc: assemblerFunc,
+	}
+
+	// apply options
+	for _, opt := range opts {
+		opt(d)
+	}
+
+	if err := d.fetchManifests(); err != nil {
+		return nil, fmt.Errorf("failed to fetch Datadog Malicious Software Packages manifests: %w", err)
+	}
+
+	return d, nil
+}
+
+func (d *datadogMalwareCertifier) fetchManifests() error {
+	npmResp, err := d.httpClient.Get(NPM_MANIFEST_URL)
+	if err != nil {
+		return fmt.Errorf("failed to fetch NPM manifest: %w", err)
+	}
+	defer npmResp.Body.Close()
+
+	if err := json.NewDecoder(npmResp.Body).Decode(&d.npmData); err != nil {
+		return fmt.Errorf("failed to decode NPM manifest: %w", err)
+	}
+
+	pypiResp, err := d.httpClient.Get(PYPI_MANIFEST_URL)
+	if err != nil {
+		return fmt.Errorf("failed to fetch PyPI manifest: %w", err)
+	}
+	defer pypiResp.Body.Close()
+
+	if err := json.NewDecoder(pypiResp.Body).Decode(&d.pypiData); err != nil {
+		return fmt.Errorf("failed to decode PyPI manifest: %w", err)
+	}
+
+	return nil
+}
+
+func (d *datadogMalwareCertifier) CertifyComponent(ctx context.Context, rootComponent interface{}, docChannel chan<- *processor.Document) error {
+	logger := logging.FromContext(ctx)
+	packageNodes, ok := rootComponent.([]*root_package.PackageNode)
+	if !ok {
+		return ErrDatadogMalwareComponentTypeMismatch
+	}
+
+	predicates := &assembler.IngestPredicates{}
+	currentTime := time.Now().UTC()
+
+	for _, node := range packageNodes {
+		purl := node.Purl
+
+		pkgInput, err := helpers.PurlToPkg(purl)
+		if err != nil {
+			logger.Debugf("failed to parse purl '%s' into package: %v", purl, err)
+			continue
+		}
+
+		// determine which dataset to check based on package type
+		var maliciousVersions []string
+		switch pkgInput.Type {
+		case "npm":
+			fullName := pkgInput.Name
+			if pkgInput.Namespace != nil && *pkgInput.Namespace != "" {
+				namespace := strings.TrimPrefix(*pkgInput.Namespace, "@")
+				namespace = strings.TrimPrefix(namespace, "%40")
+				fullName = "@" + namespace + "/" + pkgInput.Name
```

Comment:
it looks like it isn't always the case that the packages start with "@" in the dataset, could we add a check here after the trim to see if the namespace had the prefix? and add the "@" only if there was a prefix trim?

Reply:
Certainly, I've added something to this effect 👍